File: 2024/08/07/98. Validate Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node, max_val, min_val):
            if not node:
                return True
            if node.val >= max_val or node.val <= min_val:
                return False
            return dfs(node.left, node.val, min_val) and dfs(node.right, max_val, node.val)
        return dfs(root, float('inf'), float('-inf'))

File: 2024/08/07/98. Validate Binary Search Tree/solution.cpp
class Solution {
public:
    bool dfs(TreeNode* node, long long max_val, long long min_val){
        if (!node){ 
            return true;
        }
        if (node->val >= max_val || node->val <= min_val){
            return false;
        }
        return dfs(node->left, node->val, min_val) && dfs(node->right, max_val, node->val);
    }
    
    bool isValidBST(TreeNode* root) {
        return dfs(root, LLONG_MAX, LLONG_MIN);
    }
};